### PR TITLE
Fix #1397: flush websocket replay queue FIFO

### LINKED ISSUE #1397: [Detail Bug] WebSocket replay queue stalls after 10 messages and can send messages out of order after reconnect
# Detail Bug Report

https://app.detail.dev/org_89974092-6629-40ef-8b02-01754f4ef520/bugs/bug_d49cc8b7-4026-4d5a-9d9d-244cefdb0f2d

# Summary
- **Context**: `src/hooks/use-websocket-transport.ts` provides a low-level WebSocket transport hook with automatic reconnection and a message queue for reliable delivery (the "replay" policy) while disconnected.
- **Bug**: The queue flushing logic only sends up to `MAX_FLUSH_BATCH_SIZE` (10) messages upon connection establishment and leaves any remaining messages in the queue indefinitely. Additionally, new messages sent while connected bypass the queue even if it still contains re-queued messages, breaking message ordering.
- **Actual vs. expected**: Messages beyond the first 10 in the queue are never sent until another reconnection occurs, and new messages are sent immediately even if older messages are still in the queue.
- **Impact**: Critical failure of reliable message delivery and FIFO ordering. Up to 90 messages (out of `MAX_QUEUE_SIZE` = 100) can be stuck in the client queue indefinitely, and message ordering is compromised during reconnection.

# Code with Bug
```ts
// Send queued messages in batches
const toSend = validMessages.slice(0, MAX_FLUSH_BATCH_SIZE); // <-- BUG 🔴 Only sends first 10
const remaining = validMessages.slice(MAX_FLUSH_BATCH_SIZE);

for (const msg of toSend) {
  ws.send(JSON.stringify(msg));
}

// Re-queue any remaining messages
if (remaining.length > 0) {
  messageQueueRef.current = remaining; // <-- BUG 🔴 Re-queued messages never flushed again
}
```

```ts
const send = useCallback(
  (message: unknown): boolean => {
    if (wsRef.current?.readyState === WebSocket.OPEN) {
      wsRef.current.send(JSON.stringify(message)); // <-- BUG 🔴 Bypasses queue even if backlog exists
      return true;
    }
```

# Explanation
- **Queue stalling**: The `onopen` handler is the only place that drains `messageQueueRef.current`. It sends at most 10 queued messages and re-queues the rest, but never schedules/loops to flush subsequent batches. As a result, any messages beyond the first 10 remain in the queue until the next reconnect (which may not happen).
- **Ordering violation**: Because `send()` transmits immediately whenever the socket is `OPEN`, any messages sent after reconnect can jump ahead of previously queued (but not yet flushed) messages.

## Codebase Inconsistency
- `src/components/chat/use-chat-websocket.ts` calls `handleConnected` on reconnect and immediately sends a `load_session` message. Since the transport is `OPEN`, this `load_session` send bypasses the queue and can be delivered before older queued messages that were re-queued after the first 10-message flush.

# Recommended Fix
- Flush the entire queue on `onopen` (or, if batching is required, schedule continued flushing until the queue is empty).
- Preserve FIFO ordering by having `send()` enqueue when the socket is `OPEN` but `messageQueueRef.current` is non-empty (only send immediately when the queue is empty).

# History
This bug was introduced in commit aafb7f29a. This commit refactored the monolithic chat WebSocket logic into a shared `useWebSocketTransport` hook, where it introduced a 10-message batching limit during reconnection but failed to schedule subsequent flushes for remaining messages or check the queue in the `send` function, leading to permanent message stalls and out-of-order delivery.

### PR DESCRIPTION
**Pull request by martin-purplefish**
## Summary
- Fix replay-mode queue draining so reconnect flushes all queued messages, not just the first batch.
- Preserve FIFO ordering by queueing sends when a replay backlog exists, even if the socket is currently open.
- Add regression tests that cover both large backlog replay and reconnect-time ordering.

## Changes
- **WebSocket transport replay drain**: Added a dedicated queue flush path that filters stale message types, sends in bounded batches, and keeps unsent messages queued if sending fails mid-drain.
- **WebSocket transport send path**: Updated `send()` to avoid bypassing the replay queue when backlog exists, then opportunistically flush the queue when open.
- **Regression coverage**: Added jsdom hook tests for replay queues over 10 messages and for reconnect ordering with `onConnected` sends.

## Testing
- [x] Tests pass (`pnpm test`)
- [x] Types pass (`pnpm typecheck`)
- [x] Build succeeds (`pnpm build`)
- [ ] Manual testing: Not run (covered by automated regression tests)

Closes #1397

---
🏭 Forged in [Factory Factory](https://factoryfactory.ai)

---

> [!NOTE]
> **Medium Risk**
> Changes core WebSocket message delivery semantics (queue draining, batching, and send/flush behavior), which could affect ordering or drop behavior across reconnects. Risk is mitigated by added regression tests covering large backlogs, reconnect ordering, and send failures/serialization edge cases.
> 
> **Overview**
> Fixes `useWebSocketTransport` replay-mode delivery so queued messages are drained **fully and FIFO** (not just the first `MAX_FLUSH_BATCH_SIZE`), while filtering stale `stop`/`interrupt` messages only during reconnect replay and preserving unsent items if the socket closes or `send()` throws.
> 
> Updates `send()` to **not bypass the replay backlog** when the socket is open (maintaining ordering), and to opportunistically flush the backlog; adds jsdom/Vitest regression tests for oversized backlogs, reconnect ordering with `onConnected` sends, skipping non-serializable queued messages, and ensuring fresh `stop` messages aren’t dropped during live flush retries.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 38b4e94b65bae35123d468ae8571cac18aefff1f. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/hooks/use-websocket-transport.ts [hunk 1]
```diff
@@ -19,6 +19,34 @@ const STALE_MESSAGE_TYPES = new Set(['stop', 'interrupt']);
...
+): number {
+  let index = startIndex;
+  for (; index < endIndex; index += 1) {
+    const msg = messages[index];
+    if (msg === undefined) {
+      continue;
+    }
+
+    try {
+      ws.send(JSON.stringify(msg));
```
[the remark this answers is not in the document]
Comment (martin-purplefish) [reply src/hooks/use-websocket-transport.ts]: @cubic-dev-ai - fixed as suggested. I split serialization from socket send failure handling so an unserializable queued payload is skipped instead of being re-queued forever, and added a regression test for the circular-message case.

File: src/hooks/use-websocket-transport.ts [hunk 2]
```diff
@@ -313,10 +347,15 @@ export function useWebSocketTransport(
       if (messageQueueRef.current.length < MAX_QUEUE_SIZE) {
         messageQueueRef.current.push(message);
       }
+
+      if (socketIsOpen && ws) {
+        flushQueuedMessages(ws);
+      }
```
[the remark this answers is not in the document]
Comment (martin-purplefish) [reply src/hooks/use-websocket-transport.ts]: @cursor - fixed as suggested. `flushQueuedMessages` now only drops stale `stop`/`interrupt` messages during reconnect replay, while live backlog flushes preserve freshly queued commands. I added a regression test that exercises an open-socket backlog with a fresh `stop` message.

**Comment by martin-purplefish** [comment]
@cubic-dev-ai please re-review
